driver-core/src/main/com/mongodb/client/model/Filters.java

```diff
@@ -1140,10 +1141,14 @@ public String toString() {
             }
         }
 
+        private boolean isDBRef(final BsonValue value) {
+            return value.asDocument().keySet().containsAll(DBREF_KEYS);
+        }
+
         private BsonDocument createFilter(final String fieldName, final BsonValue value) {
             if (fieldName.startsWith("$")) {
                 return new BsonDocument("$not", new BsonDocument(fieldName, value));
-            } else if (value.isDocument() || value.isRegularExpression()) {
+            } else if ((value.isDocument() && !isDBRef(value)) || value.isRegularExpression()) {
```

Comment:
I think there's a related bug lurking here.  Consider this query:

   not(eq('x', new BsonDocument('y', BsonBoolean.TRUE)))

This also generates the error reported in the bug, both in master and on this branch.  But like DBRef, it works properly if you stick a $eq between the $not and the value.  In the shell, for example:

```
MongoDB Enterprise > db.not.insert({x : {y : 1} } )
MongoDB Enterprise > db.not.insert({x : {y : 2} } )
MongoDB Enterprise > db.not.find({x : {$not : {$eq : {y : 1} } } })
{ "_id" : ObjectId("5bf48dd49e384beb544f07a0"), "x" : { "y" : 2 } }
```

So perhaps the conditional should be something like:

    ((value.isDocument() && !isDBRef(value) && hasDollarPrefixedKey(value)) || ...)

That way a value like `{y : 2}` would fall through to the $eq case.

It's definitely an edge case as it's unreliable to query against a document like this (because iirc order of keys matter), but it is something that is allowed by our query language.  

What do you think?

Reply:
Good called pushed a fix